### PR TITLE
feat: filter try/except when looking at custom code

### LINKED ISSUE #21912: Allow for try/except imports for custom code
### Feature request

When uploading a model with custom code, I wanted to try and use Flash Attention in one of the modules. However, to get around a case where people might use it on CPU, I added `try/except` block for the import. 

However, I get an error when downloading locally like `This modeling file requires the following packages that were not found in your environment` which seems to come from [this file](https://github.com/huggingface/transformers/blob/main/src/transformers/dynamic_module_utils.py#L112). 

### Motivation

I want to be able to write custom model code that allows for optional imports if possible (like FlashAttention)

### Your contribution

I could try a crack at a PR?

### PR DESCRIPTION
**Pull request by zanussbaum**
# What does this PR do?

Fixes #21912 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@sgugger

### CODE REVIEW

File: src/transformers/dynamic_module_utils.py [hunk 1]
```diff
@@ -116,6 +116,9 @@ def check_imports(filename):
     with open(filename, "r", encoding="utf-8") as f:
         content = f.read()
 
+    # filter out try/except block so in custom code we can have try/except imports
+    content = re.sub(r"\s*try\s*:\s*.+\s*except\s*:", "", content, flags=re.MULTILINE)
```
Comment (sgugger) [inline src/transformers/dynamic_module_utils.py:120]:
```suggestion
    content = re.sub(r"\s*try\s*:\s*.*?\s*except\s*:", "", content, flags=re.MULTILINE)
```
This is to make sure it matches the shortest span, so we don't suppress a lot of code if there are several try except blocks 😅

**Review by sgugger — APPROVED** [review]
Thanks for opening a PR so quickly! Left a comment.

**Comment by sgugger** [comment]
Thanks again!

**Comment by zanussbaum** [comment]
Happy to help! Thanks for the great packages!

**Pull request merged**